pkg/kube/services/services_test.go

```diff
@@ -419,3 +420,25 @@ func TestFindURLFromIngress(t *testing.T) {
 		t.Logf("test %s generated URL from Ingress %s", tc.Name, actual)
 	}
 }
+
+func TestGetUrlFromVirtualService(t *testing.T) {
+	var namespace = "default"
+	var name = "vansh"
+	virtualService := &unstructured.Unstructured{}
+	virtualService.SetUnstructuredContent(map[string]interface{}{
+		"apiVersion": "networking.istio.io/v1alpha3",
+		"kind":       "VirtualService",
+		"metadata": map[string]interface{}{
+			"name":      name,
+			"namespace": namespace,
+		},
+		"spec": map[string]interface{}{
+			"hosts": []interface{}{"testing.jenkins-x.in"},
+		},
+	})
+	url, err := services.GetUrlFromVirtualService(virtualService)
```

Comment:
What we should be testing is only `FindUrlFromVsIstio`.
It has 2 error statements, so 2 test cases for that, and one for the happy path.

Reply:
So there is already a helper function to create the lazy client, so let's start by using that: https://github.com/jenkins-x/jx-helpers/blob/59d90d42864daf29ab22772bdb9c987ad4c0dcdd/pkg/kube/clients.go#L21

Can u point me to the code where u will be using this function? I know jx codebase calls the findserviceURL at multiple places, it would be simple to change the function signature to include the dynamic client (we anyways pass the normal client as an argument).
Since that's an interface, you can then use the fake dynamic client for the unit tests.

Reply:
https://github.com/jenkins-x-plugins/jx-preview/blob/main/pkg/cmd/create/create.go#L479

Reply:
A simple fix is to just pass in dynamic client (as an interface) to the FindServiceUrl function. That way u dont have to worry abt initializing anything in ur code, and can use fake dynamic client to do unit tests. Is that clear? I can write a short snippet once I have some free time.

We can update the jx preview code to pass the dynamic client to the function when jx-helpers package version gets bumped in jx-preview (and any other package that needs it).

Reply:
jx-lighthouse also needs that I have seen one error in that too regarding this.
WIll do a little research and try a little bit then will ask one thing that is not clear to me right now.

Reply:
so in the FindServiceUrl you are saying ki we should add one more function parameter (dynamic client)?
But i need that only for virtualService case if someone is using ingress in that case i do not need a dynamic client that's why i thought it should be done if we have not found the url from service and ingress.

Reply:
>so in the FindServiceUrl you are saying ki we should add one more function parameter (dynamic client)?

Yes. Let the caller do the heavy lifting. and since it's an interface, u can do the testing easily.

Reply:
OK sir : ).